python/lsst/daf/butler/core/datastoreCacheManager.py

```diff
@@ -31,27 +31,194 @@
 
 from typing import (
     TYPE_CHECKING,
+    Dict,
+    Iterable,
+    Iterator,
+    ItemsView,
+    KeysView,
+    List,
     Optional,
     Union,
+    ValuesView,
 )
 
 from abc import ABC, abstractmethod
+from collections import defaultdict
+import atexit
+import contextlib
+import datetime
 import logging
+import os
+import shutil
 import tempfile
 
+from pydantic import BaseModel, PrivateAttr
+
 from .configSupport import processLookupConfigs
 from .config import ConfigSubset
 from ._butlerUri import ButlerURI
+from .datasets import DatasetId, DatasetRef
 
 if TYPE_CHECKING:
     from .dimensions import DimensionUniverse
-    from .datasets import DatasetType, DatasetRef
+    from .datasets import DatasetType
     from .storageClass import StorageClass
     from .configSupport import LookupKey
 
 log = logging.getLogger(__name__)
 
 
+def remove_cache_directory(directory: str) -> None:
+    """Remove the specified directory and all its contents.
+    """
+    log.debug("Removing temporary cache directory %s", directory)
+    shutil.rmtree(directory, ignore_errors=True)
+
+
+def _construct_cache_path(root: ButlerURI, ref: DatasetRef, extension: str) -> ButlerURI:
+    """Construct the full path to use for this dataset in the cache.
+
+    Parameters
+    ----------
+    ref : `DatasetRef`
+        The dataset to look up in or write to the cache.
+    extension : `str`
+        File extension to use for this file.
+
+    Returns
+    -------
+    uri : `ButlerURI`
+        URI to use for this dataset in the cache.
+    """
+    # Dataset type component is needed in the name if composite
+    # disassembly is happening since the ID is shared for all components.
+    component = ref.datasetType.component()
+    component = f"_{component}" if component else ""
+    return root.join(f"{ref.id}{component}{extension}")
+
+
+def _parse_cache_name(cached_location: str) -> Dict[str, Optional[str]]:
+    """For a given cache name, return its component parts.
+
+    Changes to ``_construct_cache_path()`` should be reflected here.
+
+    Parameters
+    ----------
+    cached_location : `str`
+        The name of the file within the cache.
+
+    Returns
+    -------
+    parsed : `dict` of `str`, `str`
+        Parsed components of the file. These include:
+        - "id": The dataset ID,
+        - "component": The name of the component (can be `None`),
+        - "extension": File extension (can be `None`).
+    """
+    root_ext = cached_location.split(".", maxsplit=1)
+    root = root_ext.pop(0)
+    ext = "." + root_ext.pop(0) if root_ext else None
+
+    parts = root.split("_")
+    id_ = parts.pop(0)
+    component = parts.pop(0) if parts else None
+    return {"id": id_, "component": component, "extension": ext}
+
+
+class CacheEntry(BaseModel):
+    """Represent an entry in the cache."""
+
+    name: str
+    """Name of the file."""
+
+    size: int
```

Comment:
I'm wondering if it isn't almost as easy to get these from the filesystem each time rather than maintaining a possibly-out-of-sync in-memory copy.

But I guess it's easier to figure out datasets by id this way.

Reply:
The file sizes can't change so they are static information. Also I was under the impression that file systems don't really like repeated calls to `stat` if the cache grows big so limiting the cache check to a `readdir` rather than a `readdir`+`stat` seemed like a reasonable thing to do.